rfcs/20190815-tfdbg-v2-callbacks.md

```diff
@@ -0,0 +1,356 @@
+# TensorFlow Debugger v2: Callbacks for Eager Execution and `tf.function`s
+
+| Status        | Proposed                                             |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Shanqing Cai (cais@google.com)                       |
+| **Sponsor**   | Alexandre Passos (apassos@google.com)                |
+| **Updated**   | 2019-08-02                                           |
+
+## Objective
+
+This RFC presents an API-level design for how to achieve debugging
+instrumentation of [eager execution](https://www.tensorflow.org/guide/eager) and
+[`tf.function`](https://www.tensorflow.org/beta/guide/autograph) in TensorFlow 2
+("TF2" hereafter). This will enable users to "hook into" the following types of
+events in TF2 with a unified API:
+
+  - Eager execution of an operation ("op" hereafter) at runtime
+  - Creation of an symbolic op at graph-construction time, i.e., when a
+    user-defined Python functions is transformed into a graph (FuncGraph) with
+    the
+    [tf.function](https://www.tensorflow.org/beta/tutorials/eager/tf_function)
+    API.
+  - Runtime execution of FuncGraphs.
+
+## Motivation
+
+Such "hooks" will allow both observation and overriding of TF op's outgoing
+tensors, including concrete `EagerTensor`s and symbolic `Tensor`s (see details
+below.) This is a foundational part of the effort to bring [TensorFlow Debugger
+(tfdbg)](https://www.tensorflow.org/guide/debugger) up-to-date with TF2's
+execution paradigm.
+
+Currently, tfdbg is compatible with only the `tf.Session` API of TensorFlow 1.x.
+However, users of TF2 have raised questions and issues that indicate needs for
+a dedicated debugger, ones that cannot be easily met by a generic Python
+debugger such as pdb.  The most common examples of such needs involve
```

Comment:
Can you clarify what is in scope given the existence of pdb? Is this designed to be a supplement that handles the specific use-cases given in this paragraph, or is it a replacement for pdb?

Reply:
Good question. I added a section to address this question. It is titled "What This Proposal Offers Beyond pdb". PTAL.

Briefly, this is not a replacement for pdb. Instead, it is a TensorFlow-specific debugging instrumentation that can supplement pdb.